mysql/resource_grant.go

```diff
@@ -16,16 +17,225 @@ import (
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 )
 
-type MySQLGrant struct {
+type ObjectT string
+
+var (
+	kProcedure ObjectT = "PROCEDURE"
+	kFunction  ObjectT = "FUNCTION"
+	kTable     ObjectT = "TABLE"
+)
+
+type MySQLGrant interface {
+	GetId() string
+	SQLGrantStatement() string
+	SQLRevokeStatement() string
+	GetUserOrRole() UserOrRole
+	GrantOption() bool
+}
+
+type MySQLGrantWithDatabase interface {
+	MySQLGrant
+	GetDatabase() string
+}
+
+type MySQLGrantWithTable interface {
+	MySQLGrantWithDatabase
+	GetTable() string
+}
+
+type MySQLGrantWithPrivileges interface {
+	MySQLGrant
+	GetPrivileges() []string
+}
+
+type PrivilegesPartiallyRevocable interface {
+	SQLPartialRevokePrivilegesStatement(privilegesToRevoke []string) string
+}
+
+type UserOrRole struct {
+	Name string
+	Host string
+}
+
+func (u UserOrRole) IDString() string {
+	if u.Host == "" {
+		return u.Name
+	}
+	return fmt.Sprintf("%s@%s", u.Name, u.Host)
+}
+
+func (u UserOrRole) SQLString() string {
+	if u.Host == "" {
+		return fmt.Sprintf("'%s'", u.Name)
+	}
+	return fmt.Sprintf("'%s'@'%s'", u.Name, u.Host)
+}
+
+func (u UserOrRole) Equals(other UserOrRole) bool {
+	if u.Name != other.Name {
+		return false
+	}
+	if (u.Host == "" || u.Host == "%") && (other.Host == "" || other.Host == "%") {
+		return true
+	}
+	return u.Host == other.Host
+}
+
+type TablePrivilegeGrant struct {
 	Database   string
 	Table      string
 	Privileges []string
+	Grant      bool
+	UserOrRole UserOrRole
+	TLSOption  string
+}
+
+func (t *TablePrivilegeGrant) GetId() string {
+	return fmt.Sprintf("%s:%s", t.UserOrRole.IDString(), t.GetDatabase())
```

Comment:
I know it was like this before, but maybe it should change eventually as this doesn't look unique. That said, terraform did not check it before, so it's maybe fine.

We still have to find out what it means to change id.

Reply:
I agree. 

One thought: The next logical step after this PR is to introduce separate resources for each grant type (table privilege, procedure, and role grant), which would likely require ID changes. It might be a good idea to just bundle those changes together, so there is less work for users.

Reply:
Agreed!